packages/block-editor/src/components/block-list/use-zoom-out-block-editing-mode.js

```diff
@@ -0,0 +1,67 @@
+/**
+ * WordPress dependencies
+ */
+import { useSelect, useDispatch } from '@wordpress/data';
+import { useEffect } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import { store as blockEditorStore } from '../../store';
+import { useBlockEditContext } from '../block-edit/context';
+
+export default function useZoomOutBlockEditingMode() {
+	const { clientId: rootClientId = '' } = useBlockEditContext();
+
+	const { unsetBlockEditingMode, setBlockEditingMode } =
+		useDispatch( blockEditorStore );
+
+	const { isZoomOutMode, sectionsContainerClientId, sectionsClientIds } =
+		useSelect( ( select ) => {
+			const {
+				__unstableGetEditorMode,
+				getSectionsContainerClientId,
+				getClientIdsOfDescendants,
+			} = select( blockEditorStore );
+
+			// TODO: We need a better API as using the post type
+			// in block editor package is not allowed.
+			const postType = select( 'core/editor' ).getCurrentPostType();
+
+			const _sectionsContainerClientId =
+				getSectionsContainerClientId( postType );
+
+			return {
+				isZoomOutMode: __unstableGetEditorMode() === 'zoom-out',
+				sectionsContainerClientId: _sectionsContainerClientId,
+				sectionsClientIds: getClientIdsOfDescendants(
+					_sectionsContainerClientId
+				),
+			};
+		}, [] );
+
+	useEffect( () => {
+		if ( isZoomOutMode ) {
+			setBlockEditingMode( rootClientId, 'disabled' );
+			setBlockEditingMode( sectionsContainerClientId, 'contentOnly' );
+			sectionsClientIds.forEach( ( clientId ) => {
+				setBlockEditingMode( clientId, 'default' );
+			} );
+
+			return () => {
+				sectionsClientIds.forEach( ( clientId ) => {
+					unsetBlockEditingMode( clientId );
+				} );
+				unsetBlockEditingMode( sectionsContainerClientId );
+				unsetBlockEditingMode( rootClientId );
+			};
```

Comment:
Fees super weird to me that you're setting block editing modes individually. Are there no alternative solutions to this? Why not have a zoom out setting and change mode selectors based on that?

Reply:
Looks like a limitation of current actions. The [`DisableNonPageContentBlocks`](https://github.com/WordPress/gutenberg/blob/4cfaab6302c862ce264247ebef31b45729f04343/packages/editor/src/components/provider/disable-non-page-content-blocks.js#L48-L64) uses loops as well.

Reply:
What I mean is: why do we use actions? All this state should be derived from a single setting?

Reply:
Yes I agree.